xdsl/dialects/experimental/dmp.py

```diff
@@ -0,0 +1,455 @@
+"""
+A dialect for handling distributed memory parallelism.
+
+This is xDSL only for now.
+
+This dialects aims to provide the tools necessary to facilitate the creation
+and lowering of stencil (and other) computations in a manner that
+makes them run on node clusters.
+"""
+from __future__ import annotations
+
+from math import prod
+from typing import Sequence, ClassVar
+
+from xdsl.printer import Printer
+from xdsl.parser import Parser
+from xdsl.utils.hints import isa
+from xdsl.dialects.experimental import stencil
+from xdsl.ir import Operation, SSAValue, ParametrizedAttribute, Attribute, Dialect
+from xdsl.irdl import (
+    OptOpAttr,
+    Annotated,
+    Operand,
+    irdl_op_definition,
+    irdl_attr_definition,
+    ParameterDef,
+    IRDLOperation,
+    OpAttr,
+)
+from xdsl.dialects import builtin, memref
+
+
+@irdl_attr_definition
+class HaloExchangeDecl(ParametrizedAttribute):
+    """
+    This declares a region to be "halo-exchanged".
+    The semantics define that the region specified by offset and size
+    is the *received part*. To get the section that should be sent,
+    use the source_area() method to get the source area.
+
+     - offset gives the coordinates from the origin of the stencil field.
+     - size gives the size of the buffer to be exchanged.
+     - source_offset gives a translation (n-d offset) where the data should be
+       read from that is exchanged with the other node.
+     - neighbor gives the offset in rank to the node this data is to be
+       exchanged with
+
+    Example:
+
+        offset = [4, 0]
+        size   = [10, 1]
+        source_offset = [0, 1]
+        neighbor = -1
+
+    To visualize:
+    0   4         14
+        xxxxxxxxxx    0
+        oooooooooo    1
+
+    Where `x` signifies the area that should be received,
+    and `o` the area that should be read from.
+
+    This data will be exchanged with the node of rank (my_rank -1)
+    """
+
+    name = "dmp.exchange_decl"
+
+    offset_: ParameterDef[builtin.DenseArrayBase]
```

Comment:
why the trailing underscore?

Reply:
These are the irdl properties, the op exposes an interface to the "user" (mostly rewriters) to access these values in a more accessible manner (as tuples of ints instead of ArrayAttrs)